### ISSUE TITLE
OOM when running test.py

### ISSUE THREAD

**Issue by zhiwen-t** [WENGSYX/CMIVQA_Baseline#12]
We have finished model training, but encounterd an OOM issue when running test.py even if the batch size is set to 1.

We have identified that OOM errors happens within the for loop below.  We have tried the following methods:

- Detach output['logits'] and move it to cpu before further processing
- Stop recording ps and pe

But they do not work.  The GPU we use have 24G memory. 

Can you help take a look? thanks

https://github.com/WENGSYX/CMIVQA_Baseline/blob/ff1e54c8272a399100afc36537813cc85f426a47/task23/test.py#L129-L140

**Comment by WENGSYX** [comment]
I sincerely apologize for the difficulties and delays in addressing the issues over the past ten days. My attendance at the ICLR 2024 conference has significantly limited my availability to handle these matters promptly.

Regarding the Out of Memory (OOM) problem, it has been identified that the issue arises from the accumulation of tensors and the excessive length of some test samples. This problem persists even with a batch size of 1, leading to memory overloads.

To address this, the following solutions are recommended:

- Skip samples exceeding a predefined maximum length and set their predictions to zero.
- Perform computations on the CPU for samples that exceed the predefined maximum length.


### Solution 1: Skip Samples Exceeding Maximum Length

To implement this solution, you need to check the length of each sample's input and skip those that exceed a predefined maximum length. Here is how you can do it:

```python
# Define the maximum allowed length
MAX_LEN = 4096

def collate_fn_test(data):
    p_input_ids, p_attention_mask, token_types, start_labels, end_labels, video_ids, video_features = [], [], [], [], [], [], []
    for x in data:
        nums = 0
        for video in visual.keys():
            vi = visual[video][0]
            video_features.append(vi)
            input_id, attention = [tokenizer.cls_token_id], []
            if video in subtitle:
                if len(subtitle[video]) <= 2:
                    continue
                sub = subtitle[video]
                text = x['question']
                text = tokenizer(text)
                input_id.extend(text.input_ids)
                token_type = [0] * (len(input_id) - 1) + [1]
                for s in range(len(sub)):
                    ids = tokenizer(sub[s]['text']).input_ids[1:]
                    token_type.extend([0] * (len(ids) - 1))
                    token_type.extend([1])
                    input_id.extend(ids)
                
                # Skip samples that exceed the maximum length
                if len(input_id) > MAX_LEN:
                    p_input_ids.append([0])
                    p_attention_mask.append([0])
                    token_types.append([0])
                    video_features.append([0])
                    continue
                
                attention = [1] * len(input_id)
                p_input_ids.append(input_id)
                p_attention_mask.append(attention)
                token_types.append(token_type)
    
    vfeats = video_features
    vfeats = torch.tensor(vfeats)
    vfeats_mask = torch.tensor([[1] * vfl + [0] * (768 - vfl) for vfl in [786] * vfeats.size(0)])
    p_input_ids = [torch.tensor(p_input_id) for p_input_id in p_input_ids]
    p_attention_mask = [torch.tensor(p_attention) for p_attention in p_attention_mask]
    token_types = [torch.tensor(token_type) for token_type in token_types]
    return p_input_ids, p_attention_mask, token_types, vfeats, vfeats_mask
```

### Solution 2: Use CPU for Samples Exceeding Maximum Length

To handle large samples on the CPU, modify the `test_model` function to move these specific computations to the CPU.

```python
# Define the maximum allowed length
MAX_LEN = 4096

def test_model(model, val_loader):  # 验证
    model.eval()
    video_logits = []
    outputs = []
    tk = tqdm(val_loader, total=len(val_loader), position=0, leave=True)
    for step, (input_ids, attention_mask, token_types, vfeats, vfeats_mask) in enumerate(tk):
        logits = []
        ps, pe = [], []
        ns_dict = {}
        ops = []
        ns = 0
        for i in range(len(input_ids)):
            # Check if the sample exceeds the maximum length
            if input_ids[i].size(0) > MAX_LEN:
                device = torch.device('cpu')
            else:
                device = model.device

            with torch.no_grad():
                input_ids[i] = input_ids[i].to(device)
                attention_mask[i] = attention_mask[i].to(device)
                token_types[i] = token_types[i].to(device)
                vfeats[i] = vfeats[i].to(device)
                vfeats_mask[i] = vfeats_mask[i].to(device)

                output = model.forward_test(input_ids=input_ids[i].unsqueeze(dim=0),
                                            attention_mask=attention_mask[i].unsqueeze(dim=0),
                                            token_types=token_types[i].unsqueeze(dim=0), vfeats=vfeats[i].unsqueeze(dim=0), vfeats_mask=vfeats_mask[i].unsqueeze(dim=0)
                                            )
                ls = output['logits'].view(-1)
                for ln in range(len(ls)):
                    ns_dict[ns] = [i, int(ln / output['logits'].size(0)), int(ln % output['logits'].size(0))]
                    ns += 1
                logits.extend(ls)
                video_logits.append(output['logits'][output['start'], output['end']])
                ps.append(output['start'])
                pe.append(output['end'])
        
        a, b = torch.stack(logits).sort(descending=True)
        for n in b[:50]:
            ops.append([video_ids[ns_dict[n.item()][0]]] + get_time(ns_dict[n.item()][1], ns_dict[n.item()][2], input_ids[ns_dict[n.item()][0]], video_ids[ns_dict[n.item()][0]]))
        outputs.append({'id': step, 'question': valid[step]['question'], 'output': ops})

    return outputs
```
